src/main/java/org/apache/commons/collections4/deque/BlockingDeque.java

```diff
@@ -0,0 +1,525 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.collections4.deque;
+
+import java.io.Serializable;
+
+import java.util.AbstractCollection;
+import java.util.Collection;
+import java.util.Iterator;
+
+import java.util.Deque;
+import java.util.concurrent.locks.Condition;
+
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Decorates another {@link Deque} to support blocking operations that wait for the deque
+ * to become non-empty when retrieving an element, and wait for space to become available
+ * in the deque when storing an element if the decorated deque has a capacity limit.
+ * <p>
+ * This deque exists to support blocking operations for the decorated deque. It is thread
+ * safe.
+ * </p>
+ *
+ * @param <E> the type of elements held in this deque
+ * @since 4.5
+ */
+public class BlockingDeque<E> extends AbstractCollection<E> implements Deque<E>, Serializable {
```

Comment:
There is already a `BlockingDeque` interface in the Java JDK. 

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingDeque.html